osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/MigratePlaylistScoresToSoloScoresCommand.cs

```diff
@@ -0,0 +1,190 @@
+using osu.Game.Scoring;
+using osu.Server.Queues.ScoreStatisticsProcessor.Helpers;
+using osu.Server.Queues.ScoreStatisticsProcessor.Models;
+
+namespace osu.Server.Queues.ScoreStatisticsProcessor.Commands.Maintenance
+{
+    [Command("migrate-playlist-scores", Description = "Migrate scores from `multiplayer_scores` to `solo_scores`.")]
+    public class MigratePlaylistScoresToSoloScoresCommand : BaseCommand
+    {
+        /// <summary>
+        /// The playlist room ID to reprocess.
+        /// </summary>
+        [Required]
+        [Argument(0, Description = "Command separated list of playlist room IDs to reprocess.")]
+        public string PlaylistIds { get; set; } = string.Empty;
+
+        public async Task<int> OnExecuteAsync(CancellationToken cancellationToken)
+        {
+            foreach (string id in PlaylistIds.Split(','))
+            {
+                if (cancellationToken.IsCancellationRequested)
+                    break;
+
+                using (var db = Queue.GetDatabaseConnection())
+                {
+                    var playlistItems = await db.QueryAsync<MultiplayerPlaylistItem>("SELECT * FROM multiplayer_playlist_items WHERE room_id = @PlaylistId", new
+                    {
+                        PlaylistId = int.Parse(id),
+                    });
+
+                    foreach (var item in playlistItems)
+                    {
+                        if (cancellationToken.IsCancellationRequested)
+                            break;
+
+                        MultiplayerScore[] scores = (await db.QueryAsync<MultiplayerScore>($"SELECT * FROM multiplayer_scores WHERE playlist_item_id = {item.id}")).ToArray();
+
+                        foreach (var score in scores)
+                        {
+                            Console.WriteLine($"Reprocessing score {score.id} from playlist {item.id}");
+
+                            Ruleset ruleset = LegacyRulesetHelper.GetRulesetFromLegacyId(item.ruleset_id);
+                            HitResult maxRulesetJudgement = ruleset.GetHitResults().First().result;
+                            Dictionary<HitResult, int> statistics = JsonConvert.DeserializeObject<Dictionary<HitResult, int>>(score.statistics)!;
+                            List<HitResult> allHits = statistics
+                                                      .SelectMany(kvp => Enumerable.Repeat(kvp.Key, kvp.Value))
+                                                      .ToList();
+                            var maximumStatistics = new Dictionary<HitResult, int>();
+
+                            foreach (var groupedStats in allHits
+                                                         .Select(r => getMaxJudgementFor(r, maxRulesetJudgement))
+                                                         .GroupBy(r => r))
+                            {
+                                maximumStatistics[groupedStats.Key] = groupedStats.Count();
+                            }
```

Comment:
Not sure how much speed matters here, but I do find this implementation and the way it materialises `allHits` a bit wasteful, when instead you could do

```csharp
foreach (var resultCount in statistics)
{
    var maxResult = getMaxJudgementFor(resultCount.Key, maxRulesetJudgement);

    if (!statistics.ContainsKey(maxResult))
        statistics[maxResult] = resultCount.Value;
    else
        statistics[maxResult] += resultCount.Value;
}
```

Both _work_ though, so if this works well enough for you, this is a dismissable concern.

Reply:
I think what I have reads slightly better so I'll keep it for now.